### PR TITLE
Update `extract_values` API and add new `split_values` filter

### PR DESCRIPTION
**Pull request by user27182**
### Overview
Follow up from #5963.

In this PR:

- Update/modify the API for `extract_values` to make it more general and more explicit 
- Enable extracting values from multi-component scalar arrays
- Add a new `split_values` filter (direct alternative to `split_labels` filter originally proposed in #5950)

These features are motivated by #5968 to enable extracting surfaces generated by the `contour_labels` filter, which outputs a 2-component scalar array.

### Details
- Add `component_mode` parameter with options for extracting values using a single component,  `'any'` component, `'all'` components, or specific `'multi'` component
- Separate `values` parameter into `values` and `ranges`. This makes things more explicit and allows setting a range directly with `[lower, upper]` instead of requiring `[[lower, upper]]`
- `values` now supports specifying multi-component vectors when `component_mode`  is `values`
- Separate `keep_original_ids` param into `pass_point_ids` and `pass_cell_ids` parameters. This is more consistent with other filters.
- Make `include_cells` param dynamic to allow for use with PointSets
- Remove default behaviour of extracting unique values and splitting the output when no input values are specified.
  - ``split=True` must now be set explicitly to split meshes
  - Empty input args are no longer accepted
  - This default behaviour is what `split_values` does instead.

### CODE REVIEW

**Comment by user27182** [comment]
@pyvista-bot preview

File: pyvista/core/filters/data_set.py [hunk 1]
```diff
@@ -5284,6 +5288,18 @@ def extract_values(
...
+        component : int | 'any' | 'all', default: 'any'
+            Specify the scalar component(s) to use when the input scalar array is
+            two-dimensional. Has no effect when the scalars are one-dimensional.
+            Set this value to:
+
+            - an integer (0-indexed): extract points or cells where the specified
+              component has the specified value(s), or
+            - ``'any'``: extract points or cells where any component has the
+              specified value(s), or
+            - ``'all'`` to extract points or cells where all components have the
```
Comment (user27182) [inline pyvista/core/filters/data_set.py:5300]:
```suggestion
            - ``'all'``: extract points or cells where all components have the
```

**Comment by user27182** [comment]
LGTM

**Review by tkoyama010 — CHANGES_REQUESTED** [review]
We should avoid using if statements in test functions. If necessary, please split the function into multiple functions. It makes it easy to review :)

File: tests/core/test_dataset_filters.py [hunk 2]
```diff
@@ -2326,27 +2326,79 @@ def test_extract_values_open_intervals(grid4x4):
...
+    cell_data = labeled_data.cell_data['RegionId']
+    cell_data = np.vstack([cell_data, cell_data + offset]).T
+    labeled_data.cell_data['RegionId'] = cell_data
+
+    return small_box, big_box, labeled_data
+
+
+@pytest.mark.parametrize('component_offset', [0, -0.5, None])
+@pytest.mark.parametrize('component', [0, 1, 'any', 'all'])
+def test_extract_values_component_split(labeled_data, component_offset, component):
```
Comment (tkoyama010) [inline tests/core/test_dataset_filters.py:2357]: https://github.com/pyvista/pyvista/pull/6001#pullrequestreview-2033197591 is saying this function.

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

**Comment by user27182** [comment]
@pyvista-bot preview

File: tests/core/test_dataset_filters.py (hunk 2 above)
Comment (user27182) [reply tests/core/test_dataset_filters.py]: This test has been updated to remove the if-else blocks. if-else has been replaced with explicit test case tuples which map the inputs to the expected outputs.

**Comment by user27182** [comment]
@pyvista-bot preview

File: pyvista/core/filters/data_set.py [hunk 3]
```diff
@@ -5116,44 +5118,221 @@ def extract_points(self, ind, adjacent_cells=True, include_cells=True, progress_
...
+        .. versionadded:: 0.44
+
+        Parameters
+        ----------
+        values : number | array_like | dict, optional
+            Value(s) to extract. Can be a number, an iterable of numbers, or dictionary
+            with numeric entries. For ``dict`` inputs, either its keys or values may be
+            numeric, and the other field must be strings. The numeric field is used as
+            the input for this parameter, and if ``split`` is ``True``, the string field
+            is used to set the block names of the returned :class:`~pyvista.MultiBlock`.
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:5160]:
```suggestion
            Value(s) to extract. Can be a number, an iterable of numbers, or a dictionary
            with numeric entries. For ``dict`` inputs, either its keys or values may be
            numeric, and the other field must be strings. The numeric field is used as
            the input for this parameter, and if ``split`` is ``True``, the string field
            is used to set the block names of the returned :class:`~pyvista.MultiBlock`.
```

File: pyvista/core/filters/data_set.py [hunk 4]
```diff
@@ -5116,44 +5118,221 @@ def extract_points(self, ind, adjacent_cells=True, include_cells=True, progress_
...
+
+            .. note::
+                When extracting multi-component values with ``component_mode=multi``,
+                each value is specified as a multi-component scalar. In this case,
+                ``values`` can be a single vector, or an array of row vectors.
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:5165]:
```suggestion
                When extracting multi-component values with ``component_mode=multi``,
                each value is specified as a multi-component scalar. In this case,
                ``values`` can be a single vector or an array of row vectors.
```

File: pyvista/core/filters/data_set.py [hunk 5]
```diff
@@ -5116,44 +5118,221 @@ def extract_points(self, ind, adjacent_cells=True, include_cells=True, progress_
...
+
+        ranges : array_like | dict, optional
+            Range(s) of values to extract. Can be a single range (i.e. a sequence of
+            two numbers in the form ``[lower, upper]``), a sequence of ranges, or a
+            dictionary with range entries. Any combination of ``values`` and ``ranges``
+            may be specified together. The end points of the ranges are included in the
+            extraction. Ranges cannot be set when ``component_mode=multi``.
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:5172]:
```suggestion
            Range(s) of values to extract. Can be a single range (i.e. a sequence of
            two numbers in the form ``[lower, upper]``), a sequence of ranges, or a
            dictionary with range entries. Any combination of ``values`` and ``ranges``
            may be specified together. The endpoints of the ranges are included in the
            extraction. Ranges cannot be set when ``component_mode=multi``.
```

File: pyvista/core/filters/data_set.py [hunk 6]
```diff
@@ -5116,44 +5118,221 @@ def extract_points(self, ind, adjacent_cells=True, include_cells=True, progress_
...
+
+        component_mode : int | 'any' | 'all' | 'multi', default: 'all'
+            Specify the component(s) to use when ``scalars`` is a multi-component array.
+            Has no effect when the scalars have a single component. Must be one of:
+
+            - number: specify the component number as a 0-indexed integer. The selected
+              component must have the specified value(s).
+            - ``'any'``: any single component can have the specified value(s).
+            - ``'all'``: all individual components must have the specified values(s).
+            - ``'multi'``: the entire multicomponent item must have the specified value.
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:5200]:
```suggestion
            - number: specify the component number as a 0-indexed integer. The selected
              component must have the specified value(s).
            - ``'any'``: any single component can have the specified value(s).
            - ``'all'``: all individual components must have the specified values(s).
            - ``'multi'``: the entire multi-component item must have the specified value.
```

File: pyvista/core/filters/data_set.py [hunk 7]
```diff
@@ -5116,44 +5118,221 @@ def extract_points(self, ind, adjacent_cells=True, include_cells=True, progress_
...
+        >>> labels = dict(region1=1, region2=2, region3=3, region4=4)
+        >>>
+        >>> multiblock = image.split_values(labels)
+        >>> multiblock.keys()
+        ['region1', 'region2', 'region3', 'region4']
+
+        Plot the regions as separate meshes using the labels instead of plotting
+        the MultiBlock directly.
+
+        >>> # Clear scalar data so we can color each mesh using a single color
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:5257]:
```suggestion
        Clear scalar data so we can color each mesh using a single color
```

File: pyvista/core/filters/data_set.py [hunk 8]
```diff
@@ -5163,103 +5342,137 @@ def extract_values(
...
-            in the form ``[lower, upper]``. Multiple single-number entries and/or
-            multiple ranges may be specified. If no values are specified, each unique
-            value in the data is extracted independently and returned as a
-            :class:`~pyvista.MultiBlock` dataset (i.e. ``split`` will be set to ``True``).
+        values : number | array_like | dict, optional
+            Value(s) to extract. Can be a number, an iterable of numbers, or dictionary
+            with numeric entries. For ``dict`` inputs, either its keys or values may be
+            numeric, and the other field must be strings. The numeric field is used as
+            the input for this parameter, and if ``split`` is ``True``, the string field
+            is used to set the block names of the returned :class:`~pyvista.MultiBlock`.
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:5350]:
```suggestion
            Value(s) to extract. Can be a number, an iterable of numbers, or a dictionary
            with numeric entries. For ``dict`` inputs, either its keys or values may be
            numeric, and the other field must be strings. The numeric field is used as
            the input for this parameter, and if ``split`` is ``True``, the string field
            is used to set the block names of the returned :class:`~pyvista.MultiBlock`.
```

File: pyvista/core/filters/data_set.py [hunk 9]
```diff
@@ -5163,103 +5342,137 @@ def extract_values(
...
+
+            .. note::
+                When extracting multi-component values with ``component_mode=multi``,
+                each value is specified as a multi-component scalar. In this case,
+                ``values`` can be a single vector, or an array of row vectors.
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:5355]:
```suggestion
                When extracting multi-component values with ``component_mode=multi``,
                each value is specified as a multi-component scalar. In this case,
                ``values`` can be a single vector or an array of row vectors.
```

File: pyvista/core/filters/data_set.py [hunk 10]
```diff
@@ -5163,103 +5342,137 @@ def extract_values(
...
+
+        ranges : array_like | dict, optional
+            Range(s) of values to extract. Can be a single range (i.e. a sequence of
+            two numbers in the form ``[lower, upper]``), a sequence of ranges, or a
+            dictionary with range entries. Any combination of ``values`` and ``ranges``
+            may be specified together. The end points of the ranges are included in the
+            extraction. Ranges cannot be set when ``component_mode=multi``.
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:5362]:
```suggestion
            Range(s) of values to extract. Can be a single range (i.e. a sequence of
            two numbers in the form ``[lower, upper]``), a sequence of ranges, or a
            dictionary with range entries. Any combination of ``values`` and ``ranges``
            may be specified together. The endpoints of the ranges are included in the
            extraction. Ranges cannot be set when ``component_mode=multi``.
```

File: pyvista/core/filters/data_set.py [hunk 11]
```diff
@@ -5163,103 +5342,137 @@ def extract_values(
...
+
+        component_mode : int | 'any' | 'all' | 'multi', default: 'all'
+            Specify the component(s) to use when ``scalars`` is a multi-component array.
+            Has no effect when the scalars have a single component. Must be one of:
+
+            - number: specify the component number as a 0-indexed integer. The selected
+              component must have the specified value(s).
+            - ``'any'``: any single component can have the specified value(s).
+            - ``'all'``: all individual components must have the specified values(s).
+            - ``'multi'``: the entire multicomponent item must have the specified value.
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:5390]:
```suggestion
            When ``scalars`` is specified, this is the preferred array type to search
            for in the dataset.  Must be either ``'point'`` or ``'cell'``.

        component_mode : int | 'any' | 'all' | 'multi', default: 'all'
            Specify the component(s) to use when ``scalars`` is a multi-component array.
            Has no effect when the scalars have a single component. Must be one of:

            - number: specify the component number as a 0-indexed integer. The selected
              component must have the specified value(s).
            - ``'any'``: any single component can have the specified value(s).
            - ``'all'``: all individual components must have the specified values(s).
            - ``'multi'``: the entire multi-component item must have the specified value.
```

File: pyvista/core/filters/data_set.py [hunk 12]
```diff
@@ -5163,103 +5342,137 @@ def extract_values(
...
         adjacent_cells : bool, default: True
-            If ``True``, include cells (and their points) that contain
-            at least one of the extracted points. If ``False``, only
-            include cells which contain exclusively points from the
-            extracted points list. Has no effect if ``include_cells``
-            is ``False``. Has no effect when extracting values from
+            If ``True``, include cells (and their points) that contain at least one of
+            the extracted points. If ``False``, only include cells which contain
+            exclusively points from the extracted points list. Has no effect if
+            ``include_cells`` is ``False``. Has no effect when extracting values from
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:5400]:
```suggestion
            If ``True``, include cells (and their points) that contain at least one of
            the extracted points. If ``False``, only include cells that contain
            exclusively points from the extracted points list. Has no effect if
            ``include_cells`` is ``False``. Has no effect when extracting values from
```

File: pyvista/core/filters/data_set.py [hunk 13]
```diff
@@ -5163,103 +5342,137 @@ def extract_values(
...
         --------
-        extract_points, extract_cells, split_bodies, threshold, partition
+        split_values, extract_points, extract_cells, threshold, partition
 
         Returns
         -------
         pyvista.UnstructuredGrid or pyvista.MultiBlock
-            Extracted mesh or a composite of extracted meshes depending on
-            ``values`` and ``split``.
+            Extracted mesh or a composite of extracted meshes depending on ``split``.
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:5444]:
```suggestion
            An extracted mesh or a composite of extracted meshes, depending on ``split``.
```

File: pyvista/core/filters/data_set.py [hunk 14]
```diff
@@ -5163,103 +5342,137 @@ def extract_values(
...
-        ...     [[-np.inf, 100], [600, np.inf]]
-        ... )
+        >>> extracted = mesh.extract_values(0, include_cells=False)
+        >>> extracted.get_data_range()
+        (0.0, 0.0)
+        >>> extracted.plot(render_points_as_spheres=True, point_size=100)
+
+        Use ``ranges`` to extract a range of values from a grid's point data.
+
+        Here, we use ``+/-`` infinity to extract all values of ``100`` or less.
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:5473]:
```suggestion
        Use ``ranges`` to extract values from a grid's point data in range.

        Here, we use ``+/-`` infinity to extract all values of ``100`` or less.
```

File: pyvista/core/filters/data_set.py [hunk 15]
```diff
@@ -5278,75 +5491,293 @@ def extract_values(
...
+        >>> extracted = point_cloud.extract_values(
+        ...     values=[blue, green],
+        ...     component_mode='multi',
+        ... )
+        >>> extracted.plot(**plot_kwargs)
+
+        Use the original ids returned by the extraction to modify the original point
+        cloud.
+
+        E.g. change the color of the blue and green points to yellow.
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:5560]:
```suggestion
        Use the original IDs returned by the extraction to modify the original point
        cloud.

        For example, change the color of the blue and green points to yellow.
```

File: pyvista/core/filters/data_set.py [hunk 16]
```diff
@@ -5278,75 +5491,293 @@ def extract_values(
...
+                ):
+                    raise TypeError('Ranges must be numeric.')
+                is_valid_range = ranges_[:, 0] <= ranges_[:, 1]
+                not_valid = np.invert(is_valid_range)
+                if np.any(not_valid):
+                    invalid_ranges = ranges_[not_valid]
+                    raise ValueError(
+                        f'Invalid range {invalid_ranges[0]} specified. Lower value cannot be greater than upper value.',
+                    )
+            return values_, ranges_
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:5686]: Are we going to move these functions to the validation package?

**Review by tkoyama010 — CHANGES_REQUESTED** [review]
> - Remove default behaviour of extracting unique values and splitting the output when no input values are specified.
>   - ``split=True` must now be set explicitly to split meshes
>   - Empty input args are no longer accepted
>   - This default behaviour is what `split_values` does instead.

For the above change, I would like to see a deprecation warning output if a previous input is made. Once that and the minor modifications to the docstring only are fixed, the rest is fine. thank you for making the test code easier to read.

**Comment by user27182** [comment]
> > * Remove default behaviour of extracting unique values and splitting the output when no input values are specified.
> >   
> >   * ``split=True` must now be set explicitly to split meshes
> >   * Empty input args are no longer accepted
> >   * This default behaviour is what `split_values` does instead.
> 
> For the above change, I would like to see a deprecation warning output if a previous input is made. 

Thanks, I can add a deprecation warning. In this case a deprecation would probably be needed for other API changes, such as no longer allowing setting a range as part of `values` (this was moved to a new ``ranges`` param.).

This filter, `extract_values`, is brand new and not yet released. The deprecation would be useful only to anyone using this function directly from the main branch within the last week or so when #5963 was merged. Do you still want me to add deprecations?

**Comment by tkoyama010** [comment]
> > > * Remove default behaviour of extracting unique values and splitting the output when no input values are specified.
> > >   
> > >   * ``split=True` must now be set explicitly to split meshes
> > >   * Empty input args are no longer accepted
> > >   * This default behaviour is what `split_values` does instead.
> > 
> > 
> > For the above change, I would like to see a deprecation warning output if a previous input is made.
> 
> Thanks, I can add a deprecation warning. In this case a deprecation would probably be needed for other API changes, such as no longer allowing setting a range as part of `values` (this was moved to a new `ranges` param.).
> 
> This filter, `extract_values`, is brand new and not yet released. The deprecation would be useful only to anyone using this function directly from the main branch within the last week or so when #5963 was merged. Do you still want me to add deprecations?

Oh, I see, that's fine then. No need to do it :)

**Review by tkoyama010 — APPROVED** [review]

**Comment by user27182** [comment]
@tkoyama010 all of your other suggestions (which were marked as resolved) seem reasonable, feel free to apply them all. I didn't see them at first and so only replied to the deprecation comment initially.

**Pull request merged**